packages/server/customer-os-webhooks/service/sync_email_service.go

```diff
@@ -35,47 +44,267 @@ func NewSyncEmailService(log logger.Logger, repositories *repository.Repositorie
+			orgSyncResult, contactSyncResult, err = s.processEmail(ctx, name, email, emailData, personalEmailProviderList, source, interactionEventId)
+			if err != nil {
+				logrus.Errorf("failed to process emailData for emailData id %v :%v", emailData.Id, err)
+				return SyncResult{}, SyncResult{}, SyncResult{}, nil
+			}
+		}
+
+	} else {
+		logrus.Infof("interaction event already exists for raw emailData id %v", emailData.Id)
+		return SyncResult{}, SyncResult{}, SyncResult{}, nil
+	}
+
+	return orgSyncResult, interactionEventSyncResult, contactSyncResult, nil
+}
+
+func (s *syncEmailService) GetEmailIdForEmail(ctx context.Context, tenant string, email string, personalEmailProviderList []commonEntity.PersonalEmailProvider, source string) (string, error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "SyncEmailService.GetEmailIdForEmail")
+	defer span.Finish()
+	tracing.SetDefaultServiceSpanTags(ctx, span)
+	span.LogFields(log.String("tenant", tenant))
+	span.LogFields(log.String("email", email))
+
+	emailId, err := s.repositories.Neo4jRepositories.EmailReadRepository.GetEmailIdIfExists(ctx, tenant, email)
+	if err != nil {
+		return "", fmt.Errorf("unable to retrieve email id for tenant: %v", err)
+	}
+	if emailId != "" {
+		return emailId, nil
+	}
+
+	//if it's a personal email, we create just the email node in tenant
+	domain := utils.ExtractDomain(email)
+	for _, personalEmailProvider := range personalEmailProviderList {
+		if strings.Contains(domain, personalEmailProvider.ProviderDomain) {
+			err = s.repositories.Neo4jRepositories.EmailWriteRepository.CreateEmail(ctx, tenant, email, repository2.EmailCreateFields{
+				RawEmail:     email,
+				SourceFields: neo4jmodel.Source{Source: source},
+			})
+			if err != nil {
+				return "", fmt.Errorf("unable to create email: %v", err)
+			}
+			return email, nil
+		}
+	}
+	return emailId, nil
+}
+
+func (s *syncEmailService) processEmail(ctx context.Context, name string, email string, emailData model.EmailData, personalEmailProviderList []commonEntity.PersonalEmailProvider, source string, interactionEventId string) (SyncResult, SyncResult, error) {
+	from, to, cc, bcc, _, _ := extractEmailData(emailData)
+
+	emailId, err := s.GetEmailIdForEmail(ctx, common.GetTenantFromContext(ctx), email, personalEmailProviderList, source)
+	if err != nil {
+		logrus.Errorf("unable to retrieve emailData id for tenant: %v", err)
+		// handle error
+	}
+	if emailId == "" {
+		logrus.Errorf("unable to retrieve emailData id for tenant %s and emailData %s", common.GetTenantFromContext(ctx), email)
+		// handle error
+	}
+
+	orgSyncResult, err := s.createOrganizationDataAndSync(ctx, name, email, emailData)
+	if err != nil {
+		logrus.Errorf("unable to sync org: %v", err)
+		// handle error
+	}
+
+	contactSyncResult, err := s.createContactDataAndSync(ctx, name, email, emailData)
+	if err != nil {
+		logrus.Errorf("unable sync contact: %v", err)
+		// handle error
+	}
+
+	var eventType string
+	if email == from {
```

Comment:
this might also fail cause the email node might not be inserted in db until we actually execute this
add a wait check before with a timeout